proposals/csharp-10.0/parameterless-struct-constructors.md

```diff
@@ -2,101 +2,139 @@
 
 ## Summary
 
-Support explicit parameterless constructors for value types.
+Support explicit parameterless constructors and instance field initializers for struct types.
 
 ## Proposal
 
-If the parameterless constructor is less accessible than the value type, a warning is reported and the constructor is ignored when inaccessible.
+### Instance field initializers
+Instance field declarations for a struct may include variable initializers.
+
+As with [class field initializers](https://github.com/dotnet/csharplang/blob/master/spec/classes.md#instance-field-initialization):
+> A variable initializer for an instance field cannot reference the instance being created. 
+
+### Constructors
+A struct may declare a parameterless instance constructor.
+
+If the struct does not declare a parameterless instance constructor, and the struct has no fields with variable initializers, the struct (see [struct constructors](https://github.com/dotnet/csharplang/blob/master/spec/structs.md#constructors))
+> implicitly has a parameterless instance constructor which always returns the value that results from setting all value type fields to their default value and all reference type fields to null.
+
+If the struct does not declare a parameterless instance constructor, and the struct has field initializers, a `public` parameterless instance constructor is synthesized.
+The parameterless constructor is synthesized even if all initializer values are zeros.
+
+### Modifiers
+A parameterless instance constructor may be less accessible than the struct.
+
+A parameterless instance constructor may be declared `extern`.
+
+### Executing field initializers
+Execution of struct instance field initializers matches initialization for [classes](https://github.com/dotnet/csharplang/blob/master/spec/classes.md#instance-variable-initializers):
+> When an instance constructor has no constructor initializer, ... that constructor implicitly performs the initializations specified by the _variable_initializers_ of the instance fields ... . This corresponds to a sequence of assignments that are executed immediately upon entry to the constructor and before the implicit invocation of the direct base class constructor. The variable initializers are executed in the textual order in which they appear in the ... declaration.
+
+### Definite assignment
+Instance fields must be definitely assigned in struct instance constructors that do not have a `this()` initializer (see [struct constructors](https://github.com/dotnet/csharplang/blob/master/spec/structs.md#constructors)).
+
+Definite assignment is required for explicit parameterless constructors as well.
+Definite assignment _is not required_ for synthesized parameterless constructors.
 ```csharp
-struct NoConstructor
+struct S0
 {
+    object x = null;
+    object y;
```

Comment:
This feels wrong when compared with `S1`. We should be consistent here: if you initialize one field, you must initialize them all. Or drop the initialization requirement in `S1`.

Reply:
I've added an open issue for now so this can be discussed in LDM.

---
In reply to: [583120868](https://github.com/dotnet/csharplang/pull/4459#discussion_r583120868) [](ancestors = 583120868)